src/ai/backend/manager/types.py

```diff
@@ -82,118 +108,110 @@ class State(enum.Enum):
+    @classmethod
+    def nop(cls) -> TriState[TVal]:
+        return cls(state=_TriStateEnum.NOP, value=None)
 
-    def state(self) -> State:
-        return self._state
+    def value(self) -> TVal:
+        """
+        Returns the value of the TriState object.
+        It should only be used when the state value is unambiguously UPDATE.
+        """
+        if self._state != _TriStateEnum.UPDATE:
+            raise ValueError("Not allowed to get value when state is not UPDATE")
+        if self._value is None:
+            raise ValueError("TriState value is not set when state is UPDATE")
+        return self._value
+
+    def optional_value(self) -> Optional[TVal]:
+        """
+        Returns the value of the TriState object.
+        When state is not UPDATE, it returns None.
+        This is useful for cases where you want to check if the state is UPDATE
+        and get the value, or if it is NULLIFY or NOP and get None.
+        """
+        if self._state == _TriStateEnum.UPDATE:
+            return self._value
+        return None
 
-    def set_attr(self, obj: Any) -> None:
+    def update_dict(self, dict: dict[str, Any], attr_name: str) -> None:
         match self._state:
-            case State.UPDATE:
-                setattr(obj, self._attr_name, self._value)
-            case State.NULLIFY:
-                setattr(obj, self._attr_name, None)
-            case State.NOP:
+            case _TriStateEnum.UPDATE:
+                dict[attr_name] = self._value
+            case _TriStateEnum.NULLIFY:
+                dict[attr_name] = None
+            case _TriStateEnum.NOP:
                 pass
 
 
 class OptionalState(TriState[TVal]):
-    def __init__(self, attr_name: str, state: State, value: Optional[TVal]):
-        self._attr_name = attr_name
-        if state == State.NULLIFY:
+    """
+    OptionalState is a class that represents partial updates to an attribute of an object.
+    It is used to indicate whether an attribute should be updated or not modified at all.
+    It can be in one of two states:
+    - UPDATE: The attribute should be updated with the given value.
+    - NOP: No operation should be performed on the attribute.
+    This class is similar to TriState, but it cannot be in the NULLIFY state.
+    """
+
+    def __init__(self, state: _TriStateEnum, value: Optional[TVal]):
+        if state == _TriStateEnum.NULLIFY:
             raise ValueError("OptionalState cannot be NULLIFY")
         self._state = state
         self._value = value
 
     @classmethod
-    def update(cls, attr_name: str, value: TVal) -> OptionalState[TVal]:
-        return cls(attr_name, state=State.UPDATE, value=value)
+    def from_graphql(cls, value: Optional[TVal] | UndefinedType) -> OptionalState[TVal]:
+        if value is None:
+            raise ValueError("OptionalState cannot be NULLIFY")
+        if value is Undefined:
+            return OptionalState.nop()
+        return OptionalState.update(value)  # type: ignore
```

Comment:
```suggestion
        if isinstance(value, UndefinedType):
            return OptionalState.nop()
        return OptionalState.update(value)
```
How about checking type instead of comparing the value?
Then we don't need type ignore comment.